### ISSUE TITLE
Retraining Encoder/Generator in AAE

### ISSUE THREAD

**Issue by NicoMigenda** [zcemycl/Matlab-GAN#4]
Hello,

Iam currently implementing an AAE and found your great work! 
However, is it possible that you forgot to add the part where the Encoder/Generator is retrained? Currently the Encoder/Generator weights are only updated once with the reconstruction error from the Encoder-Decoder Part.

**Comment by zcemycl** [comment]
@NicoMigenda 
From AAE.m script, I did update encoder every iteration. 

The updates are listed in the while loop. 
From lines 56-76, 
```
        [GradEn,GradDe,GradDis] = ...
                dlfeval(@modelGradients,XBatch,...
                paramsEn,paramsDe,paramsDis,settings);

        % Update Discriminator network parameters
        [paramsDis,avgG.Dis,avgGS.Dis] = ...
            adamupdate(paramsDis, GradDis, ...
            avgG.Dis, avgGS.Dis, global_iter, ...
            settings.lrD, settings.beta1, settings.beta2);

        % Update Encoder network parameters
        [paramsEn,avgG.En,avgGS.En] = ...
            adamupdate(paramsEn, GradEn, ...
            avgG.En, avgGS.En, global_iter, ...
            settings.lrG, settings.beta1, settings.beta2);
        
        % Update Decoder network parameters
        [paramsDe,avgG.De,avgGS.De] = ...
            adamupdate(paramsDe, GradDe, ...
            avgG.De, avgGS.De, global_iter, ...
            settings.lrG, settings.beta1, settings.beta2);
```

Details are described in the modelGraidents from lines 106-125.  The generator (encoder+decoder structure) is updated together at line 123.
`[GradEn,GradDe] = dlgradient(g_loss,paramsEn,paramsDe,'RetainData',true);`

Hope this replies your question.

**Comment by NicoMigenda** [comment]
Hello,

sorry for my late response.
According to the original Paper:

Both, the adversarial network and the autoencoder are trained jointly with SGD in two phases – the reconstruction phase and the regularization phase – executed on each mini-batch. In the reconstruction phase, the autoencoder updates the encoder and the decoder to minimize the reconstruction error of the inputs. In the regularization phase, the adversarial network first updates its discriminative network to tell apart the true samples (generated using the prior) from the generated samples (the hidden codes computed by the autoencoder). The adversarial network then updates its generator (which is also the encoder of the autoencoder) to confuse the discriminative network.

From my understanding the Encoder weights have to be updated in two separated steps. Your implementation updates the Encoder weights once with a joint loss function containing the reconstruction loss and the probability term. I’m not sure if that has an impact on the training or is negotiable (or maybe I am just wrong).

Kind regards
Nico

Von: Yui Chun Leung <notifications@github.com>
Gesendet: Sonntag, 19. April 2020 12:57
An: zcemycl/Matlab-GAN <Matlab-GAN@noreply.github.com>
Cc: Migenda, Nico <nico.migenda@fh-bielefeld.de>; Mention <mention@noreply.github.com>
Betreff: Re: [zcemycl/Matlab-GAN] Retraining Encoder/Generator in AAE (#4)


@NicoMigenda<https://github.com/NicoMigenda>
From AAE.m script, I did update encoder every iteration.

The updates are listed in the while loop.
From lines 56-76,

        [GradEn,GradDe,GradDis] = ...

                dlfeval(@modelGradients,XBatch,...

                paramsEn,paramsDe,paramsDis,settings);



        % Update Discriminator network parameters

        [paramsDis,avgG.Dis,avgGS.Dis] = ...

            adamupdate(paramsDis, GradDis, ...

            avgG.Dis, avgGS.Dis, global_iter, ...

            settings.lrD, settings.beta1, settings.beta2);



        % Update Encoder network parameters

        [paramsEn,avgG.En,avgGS.En] = ...

            adamupdate(paramsEn, GradEn, ...

            avgG.En, avgGS.En, global_iter, ...

            settings.lrG, settings.beta1, settings.beta2);



        % Update Decoder network parameters

        [paramsDe,avgG.De,avgGS.De] = ...

            adamupdate(paramsDe, GradDe, ...

            avgG.De, avgGS.De, global_iter, ...

            settings.lrG, settings.beta1, settings.beta2);

Details are described in the modelGraidents from lines 106-125. The generator (encoder+decoder structure) is updated together at line 123.
[GradEn,GradDe] = dlgradient(g_loss,paramsEn,paramsDe,'RetainData',true);

Hope this replies your question.

**Comment by zcemycl** [comment]
@NicoMigenda 

Oh I see. What you mention is common for all GANs. Usually, we will not train the generator twice, but still retain that concept in mind.

For example in Keras, 
'''
# Train the discriminator d_loss_real = self.discriminator.train_on_batch(latent_real, valid) d_loss_fake = self.discriminator.train_on_batch(latent_fake, fake) d_loss = 0.5 * np.add(d_loss_real, d_loss_fake) # --------------------- # Train Generator # --------------------- # Train the generator g_loss = self.adversarial_autoencoder.train_on_batch(imgs, [imgs, valid])
'''

They also only train once. But they set discriminator trainable to false first before training generator.

In Matlab, they do it by setting generator dlgradient retaindata to true.

**Comment by NicoMigenda** [comment]
Thank you very much for that explanation. Is there further literature about that?

**Comment by zcemycl** [comment]
@NicoMigenda 
I would suggest to read how others implement vae and gan with keras and pytorch, and compare their implementations with the matlab version. 

Keras AAE:
https://github.com/eriklindernoren/Keras-GAN/blob/master/aae/aae.py
Pay attention on `discriminator.trainable` and `train_on_batch`.

PyTorch AAE:
https://github.com/eriklindernoren/PyTorch-GAN/blob/master/implementations/aae/aae.py
Pay attention on steps between `zero_grad`, `backward` and `step`.

Matlab dlgradient:
https://uk.mathworks.com/help/deeplearning/ref/dlarray.dlgradient.html

Also, compare the documentations in matlab deep learning toolbox, keras and pytorch.
